core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/ClusterBootstrapper.java

```diff
@@ -328,14 +333,19 @@ public void bootstrapNewNodes(Long stackId, Set<String> upscaleCandidateAddresse
         Set<String> clusterNodeNames = stack.getNotTerminatedInstanceMetaDataList().stream()
                 .map(InstanceMetaData::getShortHostname).collect(Collectors.toSet());
 
-        Iterator<String> iterator = recoveryHostNames.iterator();
+        List<Resource> diskResources = stack.getDiskResources();
+        Map<String, Optional<String>> imIdToFQDN = diskResources.stream()
+                .collect(Collectors.toMap(Resource::getInstanceId, volumeSet -> {
+                    Optional<VolumeSetAttributes> volumeSetAttributes = resourceAttributeUtil.getTypedAttributes(volumeSet, VolumeSetAttributes.class);
+                    return volumeSetAttributes.map(VolumeSetAttributes::getDiscoveryFQDN);
+                }));
         for (InstanceMetaData im : metaDataSet) {
             Node node = createNodeAndInitFqdnInInstanceMetadata(stack, im, clusterDomain, hostGroupNodeIndexes, clusterNodeNames);
             if (upscaleCandidateAddresses.contains(im.getPrivateIp())) {
                 // use the hostname of the node we're recovering instead of generating a new one
                 // but only when we would have generated a hostname, otherwise use the cloud provider's default mechanism
                 if (recoveredNodes && isNoneBlank(node.getHostname())) {
-                    node.setHostname(iterator.next().split("\\.")[0]);
+                    node.setHostname(imIdToFQDN.get(im.getInstanceId()).get().split("\\.")[0]);
```

Comment:
shouldn't `imIdToFQDN.get(im.getInstanceId()).isPresent()` be added to the if condition? it's strange calling get without a check first